src/app/clusters/zone-management-server/zone-management-server.cpp

```diff
@@ -0,0 +1,788 @@
+    TwoDCartesianZoneStorage twoDCartZoneStorage;
+    twoDCartZoneStorage.Set(zoneToCreate.name, zoneToCreate.use, twoDCartVertices, zoneToCreate.color);
+
+    // Call the delegate
+    status = mDelegate.CreateTwoDCartesianZone(twoDCartZoneStorage, zoneID);
+    if (status == Status::Success)
+    {
+        ZoneInformationStorage zoneInfo;
+        zoneInfo.Set(zoneID, ZoneTypeEnum::kTwoDCARTZone, ZoneSourceEnum::kUser, MakeOptional(twoDCartZoneStorage));
+
+        AddZone(zoneInfo);
+        mUserDefinedZonesCount++;
+    }
+    else
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
+        return;
+    }
+
+    response.zoneID = zoneID;
+    ctx.mCommandHandler.AddResponse(ctx.mRequestPath, response);
+}
+
+void ZoneMgmtServer::HandleUpdateTwoDCartesianZone(HandlerContext & ctx,
+                                                   const Commands::UpdateTwoDCartesianZone::DecodableType & commandData)
+{
+    uint16_t zoneID     = commandData.zoneID;
+    auto & zoneToUpdate = commandData.zone;
+
+    Status status = ValidateTwoDCartesianZone(zoneToUpdate);
+
+    if (status != Status::Success)
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
+        return;
+    }
+
+    auto foundZone =
+        std::find_if(mZones.begin(), mZones.end(), [&](const ZoneInformationStruct & z) { return z.zoneID == zoneID; });
+    if (foundZone == mZones.end())
+    {
+        ChipLogError(Zcl, "ZoneMgmt[ep=%d]: No zone exists by id %d", mEndpointId, zoneID);
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::NotFound);
+        return;
+    }
+
+    if (foundZone->zoneSource == ZoneSourceEnum::kMfg)
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError);
+        return;
+    }
+
+    if (foundZone->zoneType != ZoneTypeEnum::kTwoDCARTZone)
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::DynamicConstraintError);
+        return;
+    }
+
+    std::vector<TwoDCartesianVertexStruct> twoDCartVertices;
+
+    auto iter = zoneToUpdate.vertices.begin();
+    while (iter.Next())
+    {
+        auto vertex = iter.GetValue();
+        if (vertex.x > mTwoDCartesianMax.x || vertex.y > mTwoDCartesianMax.y)
+        {
+            ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::DynamicConstraintError);
+            return;
+        }
+        // Store the vertices
+        twoDCartVertices.push_back(vertex);
+    }
+
+    // Check the list validity
+    CHIP_ERROR err = iter.GetStatus();
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(Zcl, "HandleUpdateTwoDCartesianZone: Vertices list error: %" CHIP_ERROR_FORMAT, err.Format());
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::InvalidCommand);
+        return;
+    }
+
+    if (DoesZoneAlreadyExist(zoneToUpdate.use, twoDCartVertices))
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::AlreadyExists);
```

Comment:
We need to check zoneID first

Reply:
Added the zoneID check inside the DoesZoneAlreadyExist by passing the zoneID as a Nullable parameter.